File: dotnet/DotNetProblems/SmallProblems/Anagram.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Practicecoding.Misc
{
    class Anagram
    {
    // public static bool AreAnagrams(string firstWord, string secondWord)
    // {
    //     char [] firstWordArray = firstWord.ToCharArray();
    //     char [] secondWordArray = secondWord.ToCharArray();

    //     Array.Sort(firstWordArray);
    //     Array.Sort(secondWordArray);

    //     return new string(firstWordArray) == new string(secondWordArray);    
    // }

    public static bool AreAnagrams(string firstWord, string secondWord){
        Dictionary<char, int> firstWordCharCount = GetCharacterCount(firstWord);
        Dictionary<char, int> secondWordCharCount = GetCharacterCount(secondWord);

        if (firstWordCharCount.Count != secondWordCharCount.Count){
            return false;
        }

        else{
            foreach(var kvp in firstWordCharCount){
                if(!secondWordCharCount.ContainsKey(kvp.Key) || secondWordCharCount[kvp.Key] != kvp.Value){
                    return false;
                }
            }
            return true;
        }
    }

    public static Dictionary<char, int> GetCharacterCount(string word){
        Dictionary<char, int> characterCount = [];

        foreach(char chr in word.ToCharArray()){
            if (characterCount.ContainsKey(chr)){
                characterCount[chr] += 1;
            }
            else{
                characterCount[chr] = 1;            }
        }
        return characterCount;
    }

    public static void Init()
    {
          Console.WriteLine(AreAnagrams("listen", "silent"));
          Console.WriteLine(AreAnagrams("hello", "world"));   
    }
}

File: typescript/algos/smallproblems/anagram.ts


function getCharacterCount(input: string): {[key:string]: number}{
    let dict: {[key: string]: number } = {};

    for(const char of input) {
        if (!(char in dict)) {
            dict[char] = 1
        }
        else {
            dict[char]+=1
        }
    }
    return dict

}

function checkAnagram(first: string, second: string): boolean {
    let firstWord = getCharacterCount(first);
    let secondWord = getCharacterCount(second);

    if(firstWord.length !== secondWord.length) {
        return false
    }
    Object.keys(firstWord).forEach((key) => {
        if( secondWord[key] != firstWord[key] || !secondWord.hasOwnProperty(key)){
            return false
        }
    });
    return true;

}

console.log(`Does anagram exists for words silent and listen ? ${checkAnagram('silent', 'listen')}`)